### ISSUE TITLE
Better _draw() functionality for gradient text

### ISSUE THREAD

**Issue by choboy46** [godotengine/godot-proposals#10999]
### Describe the project you are working on

A 2D sprite based game in Godot 4.3

### Describe the problem or limitation you are having in your project

One of the previous engines I've used allowed for freely drawing with built in functions, so it delighted me to learn that Godot had its own _draw() function. However, I soon came to realize that it lacked an inherent ability to draw text with a gradient effect. Sure, I was able to figure out how to create such an effect using shaders, but that only solved part of the problem. I wanted to be able to draw multiple texts of varying colors which all had the gradient effect applied, but the way I was doing it could only allow for one gradient color per node. Moreover, only one shader effect can be applied per node at a time, and using multiple nodes to achieve my desired effect seems to be impractical.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

A possible solution without needing to change how shaders work is to add a draw function that enables users to draw a string/char with a gradient color. Users should be able to choose the direction which the gradient is drawn in. This way, you can draw multiple different colored gradient texts from a single _draw() function. Alternatively, provide a way to change shader properties in the middle of the _draw() function so that different shader properties are able to be applied to different draws.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

new draw function example:
`draw_string_gradient(font, pos, "Gradient Text", gradient1, font_size)`
`draw_string_gradient(font, pos, "Gradient Text2", gradient2, font_size)`
With a function such as this, users have a simpler way to draw text with different gradient effects applied without other draws in the same function being affected, removing the necessity to create excessive nodes.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

As I've mentioned before, you _could_ hypothetically achieve this by creating multiple nodes and give them each a shader with different uniform variables, but this would not be very practical in the context of a dialogue system.

### Is there a reason why this should be core and not an add-on in the asset library?

This could improve accessibility for people who are new to Godot, especially for those who are used to working with other engines.

**Comment by Calinou** [comment]
Are you referring to a gradient changing the letters' colors vertically (i.e. all letters are colored the same way) or horizontally (i.e. each letter has a slightly different color compared to the previous one)?

If this is the former, there's already a proposal tracking this: https://github.com/godotengine/godot-proposals/issues/2564

If you want to color each letter individually, this can already be achieved using RichTextLabel's `[color]` tag (it could be scripted so you don't have to enter them manually) or by [writing a custom RichTextEffect](https://docs.godotengine.org/en/latest/tutorials/ui/bbcode_in_richtextlabel.html#custom-bbcode-tags-and-text-effects).

**Comment by choboy46** [comment]
Vertically right now, but horizontally is another thing I look forward to. The problem that I am facing here is how font modulation works with the shaders. Normally, I would be able to change the text color mid string using a single node. However, shaders do not account for modulation, which is a problem because in order to achieve what I wanted to achieve I would need several separate duplicates of the node drawing the text in order to display different colors with the same gradient effect applied.

**Comment by bruvzg** [comment]
`RichTextEffect` only allow setting a single color for each glyph, so it won't work with gradient.

You can do text with gradient in various ways, but it will require at least one node per string/effect:
- Enable clip children property to use text as a mask (e.g, `Label` node with clip children enabled and `TextureRect` with gradient texture as a child).
- Use custom shader (e.g, `Label` + `ShaderMaterial`, I might be missing something, but seems like canvas shaders do not have transform or rect of the canvas item exposed, so you'll probably need to connect `item_rect_changed` and pass it as a custom uniform, unless you want gradient to be defined in screen/viewport space).

Both work, but can't be directly used with complex controls like RTL. 

`draw_string_gradient` seems a bit too specific, but `draw_string_with_texture` or/and an option to apply texture in the `RichTextEffect` might be useful.

**Comment by choboy46** [comment]
Small update here, I figured out that you can access the modulation color from the vertex function by saving it in a varying vec4 variable. However, I still face issues with maintaining the same gradient across all characters due to the inconsistent UVs.

**Comment by bruvzg** [comment]
> However, I still face issues with maintaining the same gradient across all characters due to the inconsistent UVs.

In case of text, `UV` is for the glyph texture atlas, you can use `SCREEN_UV` or `VERTEX` position instead:

```glsl
shader_type canvas_item;

uniform sampler2D gradient_texture;
uniform vec2 text_size;

varying vec2 vertex_coord;

void vertex() {
	vertex_coord = VERTEX / text_size;
}

void fragment() {
	COLOR = textureLod(gradient_texture, vertex_coord, 0.0) * COLOR;
}
```

<img width="349" alt="Screenshot 2024-10-21 at 07 49 12" src="https://github.com/user-attachments/assets/9fb2e217-9200-4b88-956e-5d5832f6bad5">

**Comment by choboy46** [comment]
Was actually making use of the mix() function but thanks for the advice!
```GLSL
shader_type canvas_item;

uniform bool enable = false;
uniform float fnt_ht;
uniform float vspace = 32;
uniform float mult;

const vec4 white = vec4(1.0, 1.0, 1.0, 1.0);
const vec4 dkgray = vec4(64.0 / 255.0, 64.0 / 255.0, 64.0 / 255.0 , 1.0);
const vec4 navy = vec4(0.0, 0.0, 128.0 / 255.0, 1.0);
const vec4 black = vec4(0.0, 0.0, 0.0, 1.0);

varying float vertex_mix;
varying vec4 vertex_color;
varying vec2 vertex_pass;

void vertex() {
	if (enable) {
		float vm = VERTEX.y / (fnt_ht * mult);
		float m = vspace / fnt_ht;
		vertex_mix = mod(vm, m);
		vertex_color = COLOR;
		vertex_pass = VERTEX;
	}
}

void fragment() {
	if (enable) {
		vec4 c = texture(TEXTURE, UV);
		if (c.a != 0.0) {
			vec4 gradient_color = mix(white, vertex_color, vertex_mix);
			COLOR = gradient_color;
		}
		else {
			vec2 tex_size = vec2(textureSize(TEXTURE, 0));
			if (UV.x > 0.0 && UV.y > 0.0 && texture(TEXTURE, UV - vec2(1.0 / tex_size.x, 1.0 / tex_size.y)).a != 0.0) {
				if (vertex_color == white) {
					COLOR = mix(dkgray, navy, vertex_mix);
				}
				else {
					COLOR = mix(vertex_color, black, 0.7);
				}
			}
			else {
				COLOR = vec4(0.0, 0.0, 0.0, 0.0);
			}
		}
	}
}
```

**Comment by LemmaEOF** [comment]
Shaders feel like way too much work to be able to do something like this - not to mention, they can only be applied over the *entire* text box. If you wanna do something like Deltarune's gradient effects over specific words in a dialog box, there's just no way to do that at the moment unless you write some *really* complicated shader hacks. This should really just be an engine-level feature - you should be able to set a text or outline color to be a `Gradient2D` and not have to bother with shaders (though even then that would prevent you from having separate vertical and horizontal gradients at the same time, since you don't get four-corner color control with a `Gradient2D`...)

**Comment by clayjohn** [comment]
I think this is fully solved with https://github.com/godotengine/godot/pull/90436. You can now get the actual rect data for the draw call and arbitrarily map effects over the quads used for text rendering

**Comment by LemmaEOF** [comment]
again, I would not call shaders a good solution for this! it should just be an engine-level feature, and shaders still make it hard to support applying a horizontal gradient to only individual words in a textbox!

**Comment by nebulazorua** [comment]
Also looking for a solution for something like this. I'm having to use draw_char to do per-character effects i.e shaking and thus cannot use things like LabelSettings. Shaders are also out of the question as the text shadows need to have a seperate gradient from normal text.

**Comment by sjl057** [comment]
yeah, i need this too. was trying to replicate deltarune's text and had to stop when i was drawing rectangles over text in a subviewport and using a clipping mask

i then tried to do it with just shaders, but what put it in its grave was the realization that every character would need its own canvasitem and shadermaterial, so yeah, needs to be a rendering and or textserver thing

edit: this can easily be done in nearly any other engine to my knowledge, unity with textmeshpro, gamemaker with draw_text_ext_color, and so on

**Comment by AThousandShips** [comment]
`draw_text_ext_colour` isn't really comparable to a full gradient, it only allows four control points

**Comment by sjl057** [comment]
my point was that gamemaker supports a form of it albeit limited, while
godot lacks it entirely

**Comment by Arnklit** [comment]
> I think this is fully solved with [godotengine/godot#90436](https://github.com/godotengine/godot/pull/90436). You can now get the actual rect data for the draw call and arbitrarily map effects over the quads used for text rendering

I was looking into whether there is a good way of doing gradients on text and I don't think that PR solves it. It allows you to get UVs for each individual quad and you can apply a gradient to those, but since the glyphs will be different heights (lower case, punctuation marks etc) the gradient will be uneven.

<img width="808" height="250" alt="Image" src="https://github.com/user-attachments/assets/e25eac47-ba3e-4245-bcdc-07aaf313e3fb" />

**Comment by choboy46** [comment]
Is there a way to retrieve the height per letter through this method?
